File: libs/LLMKit/Sources/LLMKit/CompletionRequest.swift
import Foundation

public enum ResponseFormatType: String, Codable, Equatable, Sendable {
    case text
    case jsonObject = "json_object"
}

public struct ResponseFormat: Codable, Equatable, Sendable {
    public var type: ResponseFormatType

    public init(type: ResponseFormatType) {
        self.type = type
    }
}

public enum ToolType: String, Codable, Equatable, Sendable {
    case function
}

public struct Tool: Codable, Equatable, Sendable {
    public var type: ToolType
    public var function: Function

    public init(type: ToolType = .function, function: Function) {
        self.type = type
        self.function = function
    }
}

public struct Function: Codable, Equatable, Sendable {
    public var description: String
    public var name: String
    public var parameters: [FunctionParameters]?

    public init(description: String, name: String, parameters: [FunctionParameters]? = nil) {
        self.description = description
        self.name = name
        self.parameters = parameters
    }
}

public struct FunctionParameters: Codable, Equatable, Sendable {
    public var type: String
    public var properties: [String: FunctionProperty]
    public var required: [String]?

    public init(type: String, properties: [String: FunctionProperty], required: [String]? = nil) {
        self.type = type
        self.properties = properties
        self.required = required
    }
}

public struct FunctionProperty: Codable, Equatable, Sendable {
    public var type: String
    public var description: String

    public init(type: String, description: String) {
        self.type = type
        self.description = description
    }
}

public enum ToolChoice: Codable, Equatable, Sendable {
    case unnamed(UnnamedToolChoice)
    case named(type: String, function: FunctionRef)

    private struct NamedToolChoice: Codable, Equatable, Sendable {
        var type: String
        var function: FunctionRef
    }

    public init(from decoder: Decoder) throws {
        let single = try decoder.singleValueContainer()
        if let unnamed = try? single.decode(UnnamedToolChoice.self) {
            self = .unnamed(unnamed)
            return
        }
        let named = try single.decode(NamedToolChoice.self)
        self = .named(type: named.type, function: named.function)
    }

    public func encode(to encoder: Encoder) throws {
        var single = encoder.singleValueContainer()
        switch self {
        case .unnamed(let unnamed):
            try single.encode(unnamed)
        case .named(let type, let function):
            try single.encode(NamedToolChoice(type: type, function: function))
        }
    }
}

public enum UnnamedToolChoice: String, Codable, Equatable, Sendable {
    case none
    case auto
    case required
}

public struct FunctionRef: Codable, Equatable, Sendable {
    public var name: String

    public init(name: String) {
        self.name = name
    }
}

public struct CompletionRequest: Codable, Equatable, Sendable {
    public var model: String
    public var messages: [Message]
    public var isStream: Bool?
    public var frequencyPenalty: Float?
    public var maxTokens: UInt32?
    public var presencePenalty: Float?
    public var responseFormat: ResponseFormat?
    public var stop: [String]?
    public var temperature: Float?
    public var topP: Float?
    public var tools: [Tool]?
    public var toolChoice: ToolChoice?
    public var logprobs: Bool?
    public var topLogprobs: UInt8?

    enum CodingKeys: String, CodingKey {
        case model
        case messages
        case isStream = "stream"
        case frequencyPenalty = "frequency_penalty"
        case maxTokens = "max_tokens"
        case presencePenalty = "presence_penalty"
        case responseFormat = "response_format"
        case stop
        case temperature
        case topP = "top_p"
        case tools
        case toolChoice = "tool_choice"
        case logprobs
        case topLogprobs = "top_logprobs"
    }

    public init(model: String, messages: [Message]) {
        self.model = model
        self.messages = messages
    }

    public func stream() -> Self {
        var copy = self
        copy.isStream = true
        return copy
    }

    public func frequencyPenalty(_ penalty: Float) -> Self {
        var copy = self
        copy.frequencyPenalty = penalty
        return copy
    }

    public func maxTokens(_ maxTokens: UInt32) -> Self {
        var copy = self
        copy.maxTokens = maxTokens
        return copy
    }

    public func presencePenalty(_ penalty: Float) -> Self {
        var copy = self
        copy.presencePenalty = penalty
        return copy
    }

    public func responseFormat(_ format: ResponseFormat) -> Self {
        var copy = self
        copy.responseFormat = format
        return copy
    }

    public func stop(_ stop: [String]) -> Self {
        var copy = self
        copy.stop = stop
        return copy
    }

    public func temperature(_ temperature: Float) -> Self {
        var copy = self
        copy.temperature = temperature
        return copy
    }

    public func topP(_ topP: Float) -> Self {
        var copy = self
        copy.topP = topP
        return copy
    }

    public func tools(_ tools: [Tool]) -> Self {
        var copy = self
        copy.tools = tools
        return copy
    }

    public func toolChoice(_ choice: ToolChoice) -> Self {
        var copy = self
        copy.toolChoice = choice
        return copy
    }

    public func toolChoice(_ choice: UnnamedToolChoice) -> Self {
        toolChoice(.unnamed(choice))
    }

    public func toolChoice(_ function: FunctionRef) -> Self {
        toolChoice(.named(type: "function", function: function))
    }

    public func logprobs(_ enabled: Bool) -> Self {
        var copy = self
        copy.logprobs = enabled
        return copy
    }

    public func topLogprobs(_ value: UInt8) -> Self {
        var copy = self
        copy.topLogprobs = value
        return copy
    }
}

public extension Function {
    func asTool() -> Tool {
        Tool(function: self)
    }
}

File: libs/llm/src/completion_request.rs
use std::collections::HashMap;

use serde::{Deserialize, Serialize};

use crate::Message;

#[derive(Clone, Default, Debug, Serialize, Deserialize)]
pub enum ResposeFormatType {
    #[default]
    #[serde(rename = "text")]
    Text,
    #[serde(rename = "json_object")]
    JsonObject,
}

#[derive(Clone, Serialize, Deserialize, Debug)]
pub struct ResponseFormat {
    r#type: ResposeFormatType,
}

#[derive(Clone, Default, Debug, Serialize, Deserialize)]
pub enum ToolType {
    #[default]
    #[serde(rename = "function")]
    Function,
}

#[derive(Clone, Default, Serialize, Debug, Deserialize)]
pub struct Tool {
    pub r#type: ToolType,
    pub function: Function,
}

#[derive(Clone, Default, Serialize, Debug, Deserialize)]
pub struct Function {
    pub description: String,
    pub name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub parameters: Option<Vec<FunctionParameters>>,
}

#[derive(Clone, Default, Serialize, Debug, Deserialize)]
pub struct FunctionParameters {
    pub r#type: String,
    pub properties: HashMap<String, FunctionProperty>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub required: Option<Vec<String>>,
}

#[derive(Clone, Serialize, Deserialize, Debug)]
pub struct FunctionProperty {
    pub r#type: String,
    pub description: String,
}

#[derive(Clone, Default, Serialize, Deserialize, Debug)]
pub struct CompletionRequest {
    pub model: String,
    pub messages: Vec<Message>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub stream: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub frequency_penalty: Option<f32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub max_tokens: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub presence_penalty: Option<f32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub response_format: Option<ResponseFormat>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub stop: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub temperature: Option<f32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub top_p: Option<f32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tools: Option<Vec<Tool>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tool_choice: Option<ToolChoice>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub logprobs: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub top_logprobs: Option<u8>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(untagged)]
pub enum ToolChoice {
    Unnamed(UnnamedToolChoice),
    Named {
        r#type: String,
        function: FunctionRef,
    },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum UnnamedToolChoice {
    None,
    Auto,
    Required,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FunctionRef {
    name: String,
}

impl CompletionRequest {
    pub fn new(model: &str, messages: &[Message]) -> Self {
        CompletionRequest {
            model: model.to_owned(),
            messages: messages.to_vec(),
            ..Default::default()
        }
    }

    pub fn stream(mut self) -> Self {
        self.stream = Some(true);
        self
    }

    pub fn frequency_penalty(mut self, penalty: f32) -> Self {
        self.frequency_penalty = Some(penalty);
        self
    }

    pub fn max_tokens(mut self, max_tokens: u32) -> Self {
        self.max_tokens = Some(max_tokens);
        self
    }

    pub fn presence_penalty(mut self, penalty: f32) -> Self {
        self.presence_penalty = Some(penalty);
        self
    }

    pub fn response_format(mut self, format: ResponseFormat) -> Self {
        self.response_format = Some(format);
        self
    }

    pub fn stop(mut self, stop: Vec<String>) -> Self {
        self.stop = Some(stop);
        self
    }

    pub fn temperature(mut self, temperature: f32) -> Self {
        self.temperature = Some(temperature);
        self
    }

    pub fn top_p(mut self, top_p: f32) -> Self {
        self.top_p = Some(top_p);
        self
    }

    pub fn tools(mut self, tools: Vec<Tool>) -> Self {
        self.tools = Some(tools);
        self
    }

    pub fn tool_choice<T: Into<ToolChoice>>(mut self, choice: T) -> Self {
        self.tool_choice = Some(choice.into());
        self
    }

    pub fn logprobs(mut self, logprobs: bool) -> Self {
        self.logprobs = Some(logprobs);
        self
    }

    pub fn top_logprobs(mut self, top_logprobs: u8) -> Self {
        self.top_logprobs = Some(top_logprobs);
        self
    }
}

impl Into<Tool> for Function {
    fn into(self) -> Tool {
        Tool {
            r#type: ToolType::Function,
            function: self,
        }
    }
}

impl Into<ToolChoice> for UnnamedToolChoice {
    fn into(self) -> ToolChoice {
        ToolChoice::Unnamed(self)
    }
}

impl Into<ToolChoice> for FunctionRef {
    fn into(self) -> ToolChoice {
        ToolChoice::Named {
            r#type: "function".to_string(),
            function: self,
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::{Message, Role, completion_request::UnnamedToolChoice};

    use super::{CompletionRequest, Function};

    #[test]
    fn tools() {
        let request = CompletionRequest::new(
            "test-model",
            &[Message {
                role: Role::User,
                content: "Hello".to_string(),
            }],
        )
        .frequency_penalty(1.0)
        .top_p(0.2)
        .tools(vec![
            Function {
                description: "Test function".to_string(),
                name: "test".to_string(),
                parameters: None,
            }
            .into(),
        ])
        .tool_choice(UnnamedToolChoice::Required);

        let json = serde_json::to_string(&request).unwrap();

        insta::assert_snapshot!(json);
    }
}
